src/test/java/one/util/streamex/TestHelpers.java

```diff
@@ -415,4 +415,85 @@ static void checkIllegalStateException(Runnable r, String key, String value1, St
                 fail("wrong exception message: " + exmsg);
         }
     }
+
+    @FunctionalInterface
+    interface Statement {
+        void evaluate() throws Throwable;
+    }
+
+    static void assertThrows(Class<? extends Throwable> expected, Statement statement) {
+        try {
+            statement.evaluate();
+            fail("Expected exception: " + expected.getName());
+        } catch (Throwable e) {
+            if (!expected.isAssignableFrom(e.getClass())) {
+                fail("Unexpected exception, " +
+                        "expected<" + expected.getName() + "> " +
+                        "but was<" + e.getClass().getName() + ">");
+            }
+        }
+    }
+
+    static <T> Collector<T, ?, ?> secondConcurrentAddAssertingCollector() {
+        return Collector.<T, Collection<T>>of(
+                SecondConcurrentAddAssertingCollection::new,
+                Collection::add,
+                (a, b) -> {
+                    throw new IllegalStateException(
+                            "Combining is not expected within secondConcurrentAddAssertingCollector");
+                },
+                Collector.Characteristics.CONCURRENT,
+                Collector.Characteristics.UNORDERED,
+                Collector.Characteristics.IDENTITY_FINISH);
+    }
+
+    private static class SecondConcurrentAddAssertingCollection<E> extends AbstractCollection<E> {

```

Comment:
Looks like a perfect job for Exchanger, see 7fabd17eb68e. It's much simpler and should test the same thing (you may check with mutation tests)

Reply:
Yes, it is quite good.